doc/source/raysgd/v2/user_guide.rst

```diff
@@ -1,167 +1,311 @@
+:ref:`backend <sgd-backends>` frameworks.
 
-Quick Start
------------
+Update training function
+~~~~~~~~~~~~~~~~~~~~~~~~
 
-RaySGD abstracts away the complexity of setting up a distributed training system. Let's take this simple example function:
+First, you'll want to update your training function to support distributed
+training.
 
-.. code-block:: python
+.. tabs::
 
-    class Net(nn.Module):
-        def __init__(self):
-            super(Net, self).__init__()
-            self.fc1 = nn.Linear(1, 128)
-            self.fc2 = nn.Linear(128, 1)
+  .. group-tab:: PyTorch
 
-        def forward(self, x):
-            x = self.fc1(x)
-            x = F.relu(x)
-            x = self.fc2(x)
-            return x
+    RaySGD will set up your distributed process group for you. You simply
+    need to add in the proper PyTorch hooks in your training function to
+    utilize it.
 
-    def train_func():
-        model = Net()
-        for x in data:
-            results = model(x)
-        return results
+    **Step 1:** Wrap your model in ``DistributedDataParallel``.
 
-To convert this to RaySGD, we add a `config` parameter to `train_func()`:
+    The `DistributedDataParallel <https://pytorch.org/docs/master/generated/torch.nn.parallel.DistributedDataParallel.html>`_
+    container will parallelize the input ``Module`` across the worker processes.
 
-.. code-block:: diff
+    .. code-block:: python
 
-    -def train_func():
-    +def train_func(config):
+        from torch.nn.parallel import DistributedDataParallel
 
-Then, we can construct the trainer function:
+        model = DistributedDataParallel(model)
 
-.. code-block:: python
+    **Step 2:** Update your ``DataLoader`` to use a ``DistributedSampler``.
 
-    from ray.util.sgd import Trainer
+    The `DistributedSampler <https://pytorch.org/docs/master/data.html#torch.utils.data.distributed.DistributedSampler>`_
+    will split the data across the workers, so each process will train on
+    only a subset of the data.
 
-    trainer = Trainer(num_workers=2)
+    .. code-block:: python
 
-Then, we can pass the function to the trainer. This will cause the trainer to start the necessary processes and execute the training function:
+        from torch.utils.data import DataLoader, DistributedSampler
 
-.. code-block:: python
+        data_loader = DataLoader(dataset,
+                                 batch_size=batch_size,
+                                 sampler=DistributedSampler(dataset))
 
-    results = trainer.run(train_func, config=None)
-    print(results)
 
-Now, let's leverage Pytorch's Distributed Data Parallel. With RaySGD, you just pass in your distributed data parallel code as as you would normally run it with `torch.distributed.launch`:
+  .. group-tab:: TensorFlow
 
-.. code-block:: python
+    .. note::
+       The current TensorFlow implementation supports
+       ``MultiWorkerMirroredStrategy`` with the Keras API. If there are other
```

Comment:
Nit: This doesn't necessarily have to be with the Keras API. It can also be with a custom training loop: https://www.tensorflow.org/tutorials/distribute/multi_worker_with_ctl

Reply:
Regular `MirroredStrategy` is also supported (not just `MultiWorkerMirroredStrategy`).